### PR TITLE
Autogenerate contribution docs from schema

### PR DESCRIPTION
**Pull request by tlambert03**
@nclack, here's rough example of how we could autogenerate nice looking markdown, while still interjecting our own prose, using jinja.  run `pip install Jinja2` then `python docs/render.py` and look in `docs/_build`

examples:

Screen Shot 2021-12-14 at 5 03 44 PM

Screen Shot 2021-12-14 at 5 04 08 PM

we can move the prose you added in the docs PR to the class docstrings and field descriptions to fill in the details.
Then, in napari, to generate the docs, it could clone this repo and run `python docs/render.py` or whatever

### CODE REVIEW

**Comment by nclack** [comment]
thanks!  I was imagining fiddling with sphinx but this is much better.  It'll be a bit before I look into it ;)

**Comment by tlambert03** [comment]
ok @nclack ... I think this is ready for a closer look.  This is more or less how I'd like the contributions page to be rendered, and it's pulling everything from the docstrings and fields themselves.  (The examples are pulled from the manifest at docs/example_manifest)

I temporarily [added the rendered result here](https://github.com/napari/npe2/blob/9cd6b1add05d2c3737bc8d16d55f79b0c21a2d58/docs/_build/contributions.md) so you can see how it ends up.

File: docs/example_manfest.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
+name: my-plugin
```
Comment (nclack) [inline docs/example_manfest.yaml:1]: thinking about renaming this to example-plugin

File: docs/render.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def main():
+    env = Environment(loader=PackageLoader("docs"), autoescape=select_autoescape())
+    _BUILD.mkdir(exist_ok=True, parents=True)
+    suffix = ".md"
+    for t in TEMPLATES.glob("*.jinja"):
+        template = env.get_template(t.name)
+        context = {"contributions": _get_contributions()}
+        (_BUILD / f"{t.stem}{suffix}").write_text(template.render(context))
```
Comment (nclack) [inline docs/render.py:98]: Two small things here:
- I'd like to rename the templates to something like `contributions.md.jinja`. This gets rid of the need for `suffix` here.
- I always appreciate seeing the modified paths get printed out.

```suggestion
def _log(path: Path) -> Path:
    print(path)
    return path


def main():
    env = Environment(loader=PackageLoader("docs"), autoescape=select_autoescape())
    _BUILD.mkdir(exist_ok=True, parents=True)
    for t in TEMPLATES.glob("*.jinja"):
        template = env.get_template(t.name)
        context = {"contributions": _get_contributions()}
        _log(_BUILD / f"{t.stem}").write_text(template.render(context))
```

**Review by nclack — COMMENTED** [review]
I'm still looking at this but some initial thoughts:

If the templates get named something like `contributions.md.jinja` then I get good syntax highlighting in vscode via the Better Jinja extension. It _also_ means that we don't need to append the suffix in render which I think makes the code a little more flexible.

I really love the example manifest. I thought "example plugin" might be better than "my plugin".

**Comment by tlambert03** [comment]
> If the templates get named something like `contributions.md.jinja` then I get good syntax highlighting in vscode via the Better Jinja extension.

lovely!  that's much "better" than the extension I'd been using: https://marketplace.visualstudio.com/items?itemName=wholroyd.jinja :)

File: docs/example_manfest.yaml (hunk 1 above)
Comment (tlambert03) [reply docs/example_manfest.yaml]: renamed

**Pull request merged**